File: smart-contracts/programs/candl/src/events.rs
use anchor_lang::prelude::*;

#[event]
pub struct MarketCreated {
    pub market: Pubkey,
    pub nft_mint: Pubkey,
    pub creator: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TradeExecuted {
    pub market: Pubkey,
    pub trader: Pubkey,
    pub is_buy: bool,
    pub sol_amount: u64,
    pub share_amount: u64,
    pub price: u64, // Price(outstanding_shares) after the trade -- see docs/03-economics.md
    pub fee_paid: u64,
    pub timestamp: i64,
}

#[event]
pub struct MarketExtended {
    pub market: Pubkey,
    pub new_expires_at: i64,
}

#[event]
pub struct MarketSettled {
    pub market: Pubkey,
    pub final_reserve: u64,
    pub timestamp: i64,
}

#[event]
pub struct SharesRedeemed {
    pub market: Pubkey,
    pub trader: Pubkey,
    pub shares: u64,
    pub sol_received: u64,
}

File: backend/src/services/indexer/events.ts
// Event shapes mirror docs/06-smart-contracts.md exactly, so decode.ts has
// a precise target to build against once the Anchor program is deployed.

export interface MarketCreatedEvent {
  type: "MarketCreated";
  market: string;
  nftMint: string;
  creator: string;
  timestamp: number;
}

export interface TradeExecutedEvent {
  type: "TradeExecuted";
  market: string;
  trader: string;
  isBuy: boolean;
  solAmount: number;
  shareAmount: number;
  price: number;
  feePaid: number;
  timestamp: number;
  signature: string;
}

export interface MarketSettledEvent {
  type: "MarketSettled";
  market: string;
  finalReserve: number;
  timestamp: number;
}

export interface MarketExtendedEvent {
  type: "MarketExtended";
  market: string;
  newExpiresAt: number;
}

export interface SharesRedeemedEvent {
  type: "SharesRedeemed";
  market: string;
  trader: string;
  shares: number;
  solReceived: number;
}

export type CandlEvent =
  | MarketCreatedEvent
  | TradeExecutedEvent
  | MarketSettledEvent
  | MarketExtendedEvent
  | SharesRedeemedEvent;
